mqtt/mqtt3/src/proto/deserialize.rs

```diff
@@ -0,0 +1,140 @@
+                write!(formatter, "an array of integers")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let topic_name = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+                let qos = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(1, &self))?;
+                let retain = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(2, &self))?;
+                let payload = seq
+                    .next_element_seed(SeededDeserializer::<Bytes>(self.0, PhantomData))?
+                    .ok_or_else(|| de::Error::invalid_length(3, &self))?;
+
+                Ok(Publication {
+                    topic_name,
+                    qos,
+                    retain,
+                    payload,
+                })
+            }
+        }
+
+        deserializer.deserialize_seq(SeededVisitor::<Self::Value>(self.0, PhantomData))
+    }
+}
+
+impl<'de, 'a> DeserializeSeed<'de> for SeededDeserializer<'a, Bytes> {
+    type Value = Bytes;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        impl<'de, 'a> Visitor<'de> for SeededVisitor<'a, Bytes> {
+            type Value = Bytes;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(formatter, "an array of integers")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let payload: Vec<u8> = seq
+                    .next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+
+                let payload = self.0.payload_de_duper.de_dupe(payload);
+
+                Ok(payload)
+            }
+        }
+
+        deserializer.deserialize_seq(SeededVisitor::<Self::Value>(self.0, PhantomData))
+    }
+}
+
+struct PayloadDeDuper {
+    loaded_bytes: HashMap<u64, Bytes>,
+}
+
+impl PayloadDeDuper {
+    fn new() -> Self {
+        Self {
+            loaded_bytes: HashMap::new(),
+        }
+    }
+
+    fn de_dupe(&mut self, bytes: Vec<u8>) -> Bytes {
+        let hash = Self::calculate_hash(&bytes);
+        if let Some(payload) = self.loaded_bytes.get(&hash) {
+            if payload == &bytes {
+                return payload.clone();
+            }
+        }
+
+        let payload = Bytes::from(bytes);
+        self.loaded_bytes.insert(hash, payload.clone());
+
+        payload
+    }
+
+    fn calculate_hash<T: Hash>(t: &T) -> u64 {
+        let mut s = DefaultHasher::new();
+        t.hash(&mut s);
+        s.finish()
+    }
+
+    fn reset(&mut self) {
```

Comment:
You can erase hash_[map|set] inside Drop implementation
```rust 
impl Drop for PayloadDeduper {
  fn drop(self) {
    self.loaded_bytes.clear();
}
```

Reply:
This actually doesn't even need this function, this is just here from when I branched from option 1.

Reply:
This code doesn't even build, I wanted to decide on an option before I spent more time on it. I just made a draft pr to make comparison easy

Reply:
Ah, I see